File: RemixedDungeon/src/main/java/com/nyrds/pixeldungeon/mobs/common/Crystal.java
package com.nyrds.pixeldungeon.mobs.common;

import static com.nyrds.pixeldungeon.levels.objects.LevelObjectsFactory.PEDESTAL;

import com.nyrds.pixeldungeon.items.chaos.ChaosCommon;
import com.nyrds.pixeldungeon.items.common.WandOfShadowbolt;
import com.nyrds.pixeldungeon.levels.objects.LevelObject;
import com.nyrds.pixeldungeon.mechanics.CommonActions;
import com.nyrds.pixeldungeon.mechanics.NamedEntityKind;
import com.watabou.pixeldungeon.Dungeon;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.actors.blobs.ConfusionGas;
import com.watabou.pixeldungeon.actors.blobs.Darkness;
import com.watabou.pixeldungeon.actors.blobs.Foliage;
import com.watabou.pixeldungeon.actors.blobs.ToxicGas;
import com.watabou.pixeldungeon.actors.buffs.Paralysis;
import com.watabou.pixeldungeon.actors.buffs.Stun;
import com.watabou.pixeldungeon.items.scrolls.ScrollOfPsionicBlast;
import com.watabou.pixeldungeon.items.wands.SimpleWand;
import com.watabou.pixeldungeon.items.wands.Wand;
import com.watabou.pixeldungeon.levels.Level;
import com.watabou.pixeldungeon.levels.Terrain;
import com.watabou.pixeldungeon.mechanics.Ballistica;
import com.watabou.pixeldungeon.scenes.GameScene;
import com.watabou.utils.Random;

import org.jetbrains.annotations.NotNull;



public class Crystal extends MultiKindMob implements IDepthAdjustable, IZapper{

	static private int ctr = 0;

	{
		movable = false;
		carcassChance = 0;
	}

	public Crystal() {

		adjustStats(Dungeon.depth);
		ensureWand(false);
	}

	static public Crystal makeShadowLordCrystal() {
		Crystal crystal = new Crystal();
		crystal.kind = 2;
		crystal.ensureWand(true);
		return crystal;
	}

	@NotNull
	private Wand ensureWand(boolean regen) {
		Wand wand = getBelongings().getItem(Wand.class);
		if (wand != null && !regen) {
			return wand;
		}

		var item = SimpleWand.createRandomSimpleWand().upgrade(Dungeon.depth/3);
		if(kind == 2 && Random.Float(1) < 0.25f) {
			item = new WandOfShadowbolt().upgrade(Dungeon.depth/2);
		}

		item.collect(this); //no treasury check, we want to keep the wand in the crystal

		return ensureWand(false);
	}

	public void adjustStats(int depth) {
		kind = (ctr++) % 2;

		hp(ht(depth * 4 + 1));

		baseDefenseSkill = depth * 2 + 1;
		baseAttackSkill = 35;
		expForKill = depth + 1;
		maxLvl = depth + 2;
		dr = expForKill /3;

		addImmunity(ScrollOfPsionicBlast.class);
		addImmunity(ToxicGas.class);
		addImmunity(Paralysis.class);
		addImmunity(Stun.class);
		addImmunity(ConfusionGas.class);
	}

	@Override
	public int getKind() {
        return super.getKind();
    }

	@Override
	public int damageRoll() {
		return Random.NormalIntRange(hp() / 2, ht() / 2);
	}

	@Override
	public boolean canAttack(@NotNull Char enemy) {
		return Ballistica.cast(getPos(), enemy.getPos(), false, true) == enemy.getPos();
	}

	@Override
	public int attackSkill(Char target) {
		if (kind < 2) {
			return 1000;
		}
		return super.attackSkill(target);
	}

	@Override
	public boolean attack(@NotNull Char enemy) {
		zap(enemy);
		return true;
	}

	@Override
	public boolean getCloser(int target) {
		return false;
	}

	@Override
    public boolean getFurther(int target) {
		return false;
	}

	@Override
	public void die(@NotNull NamedEntityKind cause) {
		int pos = getPos();

		Level level = level();

		LevelObject obj = level.getTopLevelObject(pos);

		if (obj != null && obj.getEntityKind().equals(PEDESTAL)) {
			level.remove(obj);

			level.set(pos, Terrain.EMBERS);
			int x, y;
			x = level.cellX(pos);
			y = level.cellY(pos);

			level.clearAreaFrom(Darkness.class, x - 2, y - 2, 5, 5);
			level.fillAreaWith(Foliage.class, x - 2, y - 2, 5, 5, 1);

			GameScene.updateMap();
		}
		super.die(cause);
	}

	@Override
	public boolean canBePet() {
		return false;
	}

	@Override
	protected int zapProc(@NotNull Char enemy, int damage) {
		ensureWand(false).mobWandUse(this, enemy.getPos());
		return 0;
	}

	@Override
	public void onActionTarget(String action, Char actor) {
		if(action.equals(CommonActions.MAC_STEAL)) {
			ChaosCommon.doChaosMark(getPos(), Dungeon.depth * 3);
			die(actor);
		}
	}
}

File: scripts/mobs/Crystal.lua
--[[
  Crystal (batch 17d-5, was mobs/common/Crystal.java). Immobile zapper that
  fires its held wand, free of charge, at anything in line of sight. The
  Shadow Lord spawns the kind-2 pedestal variant (25% shadowbolt wand);
  plain pedestal rooms spawn kind 0 with a random simple wand. Kind < 2
  auto-hits (attackSkill 1000, java kept). Dying on a pedestal clears the
  Darkness around it and grows Foliage. Being stolen from marks the thief
  with chaos and kills the crystal. Stats scale with depth.
]]
local RPD = require "scripts/lib/commonClasses"
local mob = require "scripts/lib/mob"

local function ensureWand(self)
	local wand = self:getBelongings():getItemPartialMatch("WandOf")
	if wand:valid() then
		return wand
	end

	local depth = RPD.Dungeon.depth
	if self:getKind() == 2 and math.random() < 0.25 then
		wand = RPD.ItemFactory:itemByName("WandOfShadowbolt")
		wand:upgrade(math.floor(depth / 2))
	else
		wand = RPD.SimpleWand:createRandomSimpleWand()
		wand:upgrade(math.floor(depth / 3))
	end

	wand:collect(self) -- no treasury check: the wand stays in the crystal
	return wand
end

return mob.init{
	spawn = function(self, level)
		ensureWand(self)
	end,

	stats = function(self)
		local depth = RPD.Dungeon.depth
		self:ht(depth * 4 + 1)
		self:setBaseDefenseSkill(depth * 2 + 1)
		self:setExpForKill(depth + 1)
		self:setMaxLvl(depth + 2)
	end,

	dr = function(self)
		return math.floor((RPD.Dungeon.depth + 1) / 3)
	end,

	attackSkill = function(self, target)
		if self:getKind() < 2 then
			return 1000
		end
		return nil
	end,

	damageRoll = function(self)
		return RPD.Random:NormalIntRange(math.floor(self:hp() / 2), math.floor(self:ht() / 2))
	end,

	canAttack = function(self, enemy)
		return RPD.Ballistica:cast(self:getPos(), enemy:getPos(), false, true) == enemy:getPos()
	end,

	doAttack = function(self, enemy)
		-- java attack() always zapped, even point blank
		if RPD.Dungeon.level:distance(self:getPos(), enemy:getPos()) <= 1 then
			self:zap(enemy)
			self:spend(self:attackDelay())
			return true
		end
		return nil
	end,

	zapProc = function(self, enemy, damage)
		ensureWand(self):mobWandUse(self, enemy:getPos())
		return 0
	end,

	die = function(self, cause)
		local level = RPD.Dungeon.level
		local pos = self:getPos()
		local obj = level:getTopLevelObject(pos)

		if obj ~= nil and obj:getEntityKind() == "pedestal" then
			level:remove(obj)
			level:set(pos, RPD.Terrain.EMBERS)
			local x, y = level:cellX(pos), level:cellY(pos)
			level:clearAreaFrom(RPD.Blobs.Darkness, x - 2, y - 2, 5, 5)
			level:fillAreaWith(RPD.Blobs.Foliage, x - 2, y - 2, 5, 5, 1)
			RPD.GameScene:updateMap()
		end
	end,

	-- direct java dispatch (CustomMob.onActionTarget): (script, the crystal, action, actor)
	onActionTarget = function(self, me, action, actor)
		if action == RPD.Actions.ch_steal then
			RPD.ChaosCommon:doChaosMark(me:getPos(), RPD.Dungeon.depth * 3)
			me:die(actor)
		end
	end,
}
